src/main/java/org/testng/internal/objects/GuiceBasedObjectDispenser.java

```diff
@@ -0,0 +1,52 @@
+package org.testng.internal.objects;
+
+import com.google.inject.Injector;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.testng.ITestContext;
+import org.testng.annotations.Guice;
+import org.testng.internal.annotations.AnnotationHelper;
+import org.testng.internal.objects.pojo.CreationAttributes;
+import org.testng.internal.objects.pojo.BasicAttributes;
+
+/**
+ * A Guice backed Object dispenser that is aware of Dependency Injection
+ */
+class GuiceBasedObjectDispenser implements IObjectDispenser {
+
+  private IObjectDispenser dispenser;
+  private static final Map<Integer, GuiceHelper> helpers = new ConcurrentHashMap<>();
+
+  @Override
+  public void setNextDispenser(IObjectDispenser dispenser) {
+    this.dispenser = dispenser;
+  }
+
+  @Override
+  public Object dispense(CreationAttributes attributes) {
+    if (attributes.getBasicAttributes() != null) {
+      BasicAttributes sa = attributes.getBasicAttributes();
+      if (cannotDispense(sa.getTestClass().getRealClass())) {
+        return this.dispenser.dispense(attributes);
+      }
+      ITestContext ctx = attributes.getContext();
+      GuiceHelper helper = helpers.computeIfAbsent(ctx.hashCode(), k -> new GuiceHelper(ctx));
+      Injector injector = helper.getInjector(sa.getTestClass(), ctx.getInjectorFactory());
```

Comment:
@krmahadevan , @juherr , this looks sad :(
1. It leaks memory in into a `static Map`
2. It might produce wrong data due to `hashCode` being used as a key

Reply:
@vlsi - I didnt understand point 1. Please elaborate.

> It might produce wrong data due to hashCode being used as a key

Ok. But unless and until a person voluntarily passes in a dummy hashcode via that JVM argument, do you mind helping me understand how this will produce wrong data ? The possibility of a hashCode collision between `ITestContext` (which represents `<test>`)<-- Trying to understand this use case.

Reply:
What is the purpose of using `hashCode` here instead of context object itself?

Reply:
I needed a unique way of identifying the `ITestContext`. So chose to use its `hashCode()` which is also _almost_ unique unless and until the collisions happen (which I understand can and will happen when there are way too many objects, the likelihood of that happening IMO is un-realistic).

Reply:
`ctx` uniquely identifies the object. Why don't you use it?

Reply:
Sure we can. But I would like to understand what is wrong with using `hashCode()` in this context ?

Reply:
> when there are way too many objects

How many?
Typical hash code has 28..31 bit entropy, so 2^15 == 32768 objects would yield a collision with a probability 50%.
It looks very possible in practice, and I would like to avoid that.

What is the purpose of baking such approaches to the core, especially when better alternatives are available?

Reply:
> But I would like to understand what is wrong with using hashCode() in this context ?

Here you go: test engine would yield wrong results (e.g. false positive or false negative test outcome), and the users won't know the reason.

It is very likely the issue would happen, and there is absolutely no reason to use `hashCode` here 🤷‍♂️

Reply:
So if I understand your explanation correctly, you are saying that if a user runs with `<test>` 32768 objects then this can occur. And you are suggesting that its very possible in practice. I kind of dont agree with that argument and I say this based on close to a decade of being associated with TestNG

Reply:
> > But I would like to understand what is wrong with using hashCode() in this context ?
> 
> Here you go: test engine would yield wrong results (e.g. false positive or false negative test outcome), and the users won't know the reason.
> 
> It is very likely the issue would happen, and there is absolutely no reason to use `hashCode` here 🤷‍♂️

TestNG would report wrong results. yes that's obvious. But the likely hood of this issue to happen that's the part wherein I still dont agree. Creating 32k `<test>` (This is not the same as a `@Test` method FYI) to me doesnt sound realistic.

Reply:
It is really sad what you do here, and it is really sad to spend time on discussing this.

What you do is *you introduce silent data corruptions to the test engine* when better alternatives exist.

Compare two approaches:
a. `helpers.computeIfAbsent(ctx.hashCode(),`.  There's a **probability of a wrong result**, and it is not clear what would happen in case of a hash code collision.
b. `helpers.computeIfAbsent(ctx,`   There's no way to get the wrong result in that case.

Now you claim that "probability is low". Why play with fire at all? Why introduce new possibilities for the bugs to the test engine at all?

What are we discussing here?

Reply:
Just in case, I got here when I tried to decouple Guice from core. No way I was trying to find `hashCode` usages in the code.

Reply:
@vlsi  - 

> It is really sad what you do here, and it is really sad to spend time on discussing this.

No reason to get all sad. It's just code and there are always ways of making it make better. 

> What are we discussing here?

This exactly what I am trying to find out. What exactly are you trying to discuss here on a merged PR which is almost 5 months old ?

Reply:
This discussion has context -- there's a specific code line.

I assumed there were reasons to use hashCode, so I asked.

If you agree this is a bug, then, here's the issue: https://github.com/cbeust/testng/issues/2580

Reply:
> I assumed there were reasons to use hashCode, so I asked.

I believe i have explained the rationale behind using it.

Reply:
@vlsi 

Your help and comments are welcome but this:

> It is really sad what you do here, and it is really sad to spend time on discussing this.

is crossing the line.

Please be civil.